### PR TITLE
Feature: email new evaluations(ID#129355383)

### PR DESCRIPTION
**Pull request by Arthurlpgc**
This change is [Reviewable](https://reviewable.io/reviews/pauloborba/ta/33)

### CODE REVIEW

File: test/cucumber/steps/EmailSteps.groovy [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+            page.fillStudentDetails(name, email, email)
+            page.selectAddStudent()
+            to CreateCriterionPage
+            page.createCriterion(criteria)
+            to AddEvaluationPage
+            page.chooseCriterion(criteria)
+            page.chooseValue(grade)
+            page.selectAddEvaluation()
+}
+Then(~/^An email is sent to “([^"]*)” Telling he received an “([^"]*)” on “([^"]*)”$/) {
```
Comment (rodCarneiro) [inline test/cucumber/steps/EmailSteps.groovy]: Não é necessário receber as duas ultimas strings já que o step não utiliza eles para verificar que o email foi enviado.

File: test/cucumber/steps/EmailSteps.groovy [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+        to StudentPage
+        page.selectStudentByEmail(email)
+        at ShowStudentPage
+        assert page.isAllSent()
+}
+When(~/^I request to send new grades$/) { ->
+    to StudentPage
+    page.sendNewEvaluations()
+}
+Given(~/^All grades on the system have been already sent$/) { ->
```
Comment (rodCarneiro) [inline test/cucumber/steps/EmailSteps.groovy:44]: o step acessa a função para enviar o email porém não verifica se ela realmente enviou os emails, poderia ser verificado embaixo de page.sendNewEvaluations() se a variável sent, de Evaluation, que indica se as evaluations foram enviadas ou não, está true

File: test/cucumber/steps/EmailSteps.groovy [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+    to CreateCriterionPage
+    page.createCriterion("C1")
+    to AddEvaluationPage
+    page.chooseCriterion("C1")
+    page.chooseValue("MPA")
+    page.selectAddEvaluation()
+    to StudentPage
+    page.sendNewEvaluations()
+}
+And(~/^I am on the students index page$/) { ->
```
Comment (rodCarneiro) [inline test/cucumber/steps/EmailSteps.groovy:57]: Como ele quer verificar se está na página index, era para utilizar o "at" ao invés de "to"

File: test/functional/pages/StudentPages/ResendPage.groovy [hunk 4]
```diff
@@ -0,0 +1,14 @@
+package pages.StudentPages
+
+import geb.Page
+
+class ResendPage extends Page {
+
+    static url = "/TA/student/resend"
+
+    static at =  {
+        title ==~ /Resend Evaluations/
```
Comment (rodCarneiro) [inline test/functional/pages/StudentPages/ResendPage.groovy:10]: o jeito certo de passar o titulo da página seria utilizando a classe GetPageTitle() para tornar-la independente de linguagem

File: test/cucumber/steps/EmailSteps.groovy [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+    at ResendPage
+}
+When(~/^I successfully import students evaluation for the "([^"]*)" criterion\.\.\.$/) { String arg1 ->
+    // Write code here that turns the phrase above into concrete actions
+    throw new PendingException()
+}
+Then(~/^The system is marked as having grades to send$/) { ->
+    // Write code here that turns the phrase above into concrete actions
+    throw new PendingException()
+}
```
Comment (rodCarneiro) [inline test/cucumber/steps/EmailSteps.groovy:62]: Tem que limpar o ambiente no fim do teste

File: test/cucumber/steps/EmailSteps.groovy [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+            page.chooseValue(grade)
+            page.selectAddEvaluation()
+}
+Then(~/^An email is sent to “([^"]*)” Telling he received an “([^"]*)” on “([^"]*)”$/) {
+    String email,String grade,String criteria->
+        to StudentPage
+        page.selectStudentByEmail(email)
+        at ShowStudentPage
+        assert page.isAllSent()
+}
```
Comment (rodCarneiro) [inline test/cucumber/steps/EmailSteps.groovy:39]: Tem que limpar o ambiente no fim do teste

**Comment by pauloborba** [comment]
Reviewed 11 of 16 files at r1.
Review status: 11 of 16 files reviewed at latest revision, 25 unresolved discussions.

---

_[grails-app/controllers/ta/StudentController.groovy, line 229 at r1](https://reviewable.io:443/reviews/pauloborba/ta/33#-KWwBIln9Eu-FLYzN7ac:-KWwBIln9Eu-FLYzN7ad:b-6m0sl7) ([raw file](https://github.com/pauloborba/ta/blob/b8dfdf0806966649aea20a09fb7124bd3fe4d00d/grails-app/controllers/ta/StudentController.groovy#L229)):_

> ``` Groovy
> 
>     def resend(){
>         render view: "resend"
> ```

nem precisa pois é o default de grails... 

---

_[grails-app/controllers/ta/StudentController.groovy, line 236 at r1](https://reviewable.io:443/reviews/pauloborba/ta/33#-KWwBoY72lciJm1syN0p:-KWwBoY72lciJm1syN0q:bmed6j2) ([raw file](https://github.com/pauloborba/ta/blob/b8dfdf0806966649aea20a09fb7124bd3fe4d00d/grails-app/controllers/ta/StudentController.groovy#L236)):_

> ``` Groovy
>         for (Student student : Student.findAll()) {
>             sentAny = sentAny || student.sendNewEvaluations()
>             savee(student)
> ```

precisa disso mesmo? não seria um student.save()?

---

_[grails-app/controllers/ta/StudentController.groovy, line 238 at r1](https://reviewable.io:443/reviews/pauloborba/ta/33#-KWwC1tSN45wtS8EkWiq:-KWwC1tSN45wtS8EkWir:bmez3o4) ([raw file](https://github.com/pauloborba/ta/blob/b8dfdf0806966649aea20a09fb7124bd3fe4d00d/grails-app/controllers/ta/StudentController.groovy#L238)):_

> ``` Groovy
>             savee(student)
>         }
>         if(sentAny) redirect action:"index", method:"GET"
> ```

não dá nenhuma mensagem de sucesso ou insucesso?

---

_[grails-app/domain/ta/EvaluationsByCriterion.groovy, line 10 at r1](https://reviewable.io:443/reviews/pauloborba/ta/33#-KWwCnyIUxmWOG9qE3y3:-KWwCnyIUxmWOG9qE3y4:b-wa6ybl) ([raw file](https://github.com/pauloborba/ta/blob/b8dfdf0806966649aea20a09fb7124bd3fe4d00d/grails-app/domain/ta/EvaluationsByCriterion.groovy#L10)):_

> ``` Groovy
>     static hasMany = [evaluations:Evaluation]
>     double criterionAverage
>     boolean hasUnsent
> ```

hasUnsentEvaluations? achei o nome esquisito... e não parece estar sendo atualizado mais abaixo

---

_[grails-app/domain/ta/EvaluationsByCriterion.groovy, line 45 at r1](https://reviewable.io:443/reviews/pauloborba/ta/33#-KWwD4QD02gtv_Jov6cf:-KWwD4QD02gtv_Jov6cg:bbs7fws) ([raw file](https://github.com/pauloborba/ta/blob/b8dfdf0806966649aea20a09fb7124bd3fe4d00d/grails-app/domain/ta/EvaluationsByCriterion.groovy#L45)):_

> ``` Groovy
> 
>     String getNewEvaluations(){
>         String s=""
> ```

melhora o espaçamento

---

_[grails-app/domain/ta/EvaluationsByCriterion.groovy, line 49 at r1](https://reviewable.io:443/reviews/pauloborba/ta/33#-KWwDg6yXd3YAQN_61LI:-KWwDg6yXd3YAQN_61LJ:b-isc6ha) ([raw file](https://github.com/pauloborba/ta/blob/b8dfdf0806966649aea20a09fb7124bd3fe4d00d/grails-app/domain/ta/EvaluationsByCriterion.groovy#L49)):_

> ``` Groovy
>             if(!ev.sent)s+=" "+ev.value+","
>             ev.sent=true
>         }
> ```

lógica confusa. não entendi porque está montando uma string aqui. no mínimo o nome do método deveria ser diferente

---

_[grails-app/domain/ta/EvaluationsByCriterion.groovy, line 52 at r1](https://reviewable.io:443/reviews/pauloborba/ta/33#-KWwDTfGVQPzkBS5NyeR:-KWwDTfH7YdEf9By_Ema:bx4i1uh) ([raw file](https://github.com/pauloborba/ta/blob/b8dfdf0806966649aea20a09fb7124bd3fe4d00d/grails-app/domain/ta/EvaluationsByCriterion.groovy#L52)):_

> ``` Groovy
>         if(s.length()==0)return ""
>         s=s.substring(0,s.length()-1)
>         return ""+criterion.getDescription()+":"+s
> ```

merecia extração de método

---

_[grails-app/domain/ta/Student.groovy, line 7 at r1](https://reviewable.io:443/reviews/pauloborba/ta/33#-KWwE8x6fOEchT7ejFyp:-KWwE8x6fOEchT7ejFyq:bn1k4hy) ([raw file](https://github.com/pauloborba/ta/blob/b8dfdf0806966649aea20a09fb7124bd3fe4d00d/grails-app/domain/ta/Student.groovy#L7)):_

> ``` Groovy
>     String login;
>     String email;
>     def messagingService;
> ```

deveria ser acessível a partir do controlador, não do modelo

---

_[grails-app/domain/ta/Student.groovy, line 58 at r1](https://reviewable.io:443/reviews/pauloborba/ta/33#-KWwEQolV-wRBFUwOwfx:-KWwEQolV-wRBFUwOwfy:bkbksiz) ([raw file](https://github.com/pauloborba/ta/blob/b8dfdf0806966649aea20a09fb7124bd3fe4d00d/grails-app/domain/ta/Student.groovy#L58)):_

> ``` Groovy
> 
>     boolean sendNewEvaluations(){
>         String message="";
> ```

merece extrações de método

---

_[grails-app/domain/ta/Student.groovy, line 59 at r1](https://reviewable.io:443/reviews/pauloborba/ta/33#-KWwEo8VMwSjdXG44ai4:-KWwEo8WSZiJFjU6TMeH:b-dtstku) ([raw file](https://github.com/pauloborba/ta/blob/b8dfdf0806966649aea20a09fb7124bd3fe4d00d/grails-app/domain/ta/Student.groovy#L59)):_

> ``` Groovy
>     boolean sendNewEvaluations(){
>         String message="";
>         this.criteriaAndEvaluations.each { EvaluationsByCriterion CAE ->
> ```

evita maiúsculas 

---

_[grails-app/domain/ta/Student.groovy, line 65 at r1](https://reviewable.io:443/reviews/pauloborba/ta/33#-KWwFGQwjWEJlyWejNt1:-KWwFGQwjWEJlyWejNt2:b3kre) ([raw file](https://github.com/pauloborba/ta/blob/b8dfdf0806966649aea20a09fb7124bd3fe4d00d/grails-app/domain/ta/Student.groovy#L65)):_

> ``` Groovy
>                 message += "\n"
>                 CAE.setHasUnsent(false)
>                 println(""+criteriaAndEvaluations.get(0).hasUnsent+CAE.getHasUnsent()+""+CAE.getNewEvaluations())
> ```

?

---

_[test/cucumber/EmailNewGrades.feature, line 8 at r1](https://reviewable.io:443/reviews/pauloborba/ta/33#-KWwFYkbTdVrVHALgBK6:-KWwFYkbTdVrVHALgBK7:bc3ffrn) ([raw file](https://github.com/pauloborba/ta/blob/b8dfdf0806966649aea20a09fb7124bd3fe4d00d/test/cucumber/EmailNewGrades.feature#L8)):_

> ``` Cucumber
>   #-Created By Arthur Costa
>   Scenario: Sending new grades
>     Given There are grades from student “Arthur” with email “alpgc@cin.ufpe.br” evaluated as “MANA” in the “Cenarios” criteria on the system that were not yet sent
> ```

Given The only not yet sent grade...  ai amarra que não tem outras notas pendentes, senão o then não estaria preciso

---

_[test/cucumber/EmailNewGrades.feature, line 10 at r1](https://reviewable.io:443/reviews/pauloborba/ta/33#-KWwG77JLksneBeQJnhV:-KWwG77JLksneBeQJnhW:bopausw) ([raw file](https://github.com/pauloborba/ta/blob/b8dfdf0806966649aea20a09fb7124bd3fe4d00d/test/cucumber/EmailNewGrades.feature#L10)):_

> ``` Cucumber
>     Given There are grades from student “Arthur” with email “alpgc@cin.ufpe.br” evaluated as “MANA” in the “Cenarios” criteria on the system that were not yet sent
>     When I request to send new grades
>     Then An email is sent to “alpgc@cin.ufpe.br” Telling he received an “MANA” on “Cenarios”
> ```

seria bom um cenário com pelo menos duas notas pendentes, já que isso é parte importante da solução

---

_[test/cucumber/EmailNewGrades.feature, line 18 at r1](https://reviewable.io:443/reviews/pauloborba/ta/33#-KWwGLPtfCHBtBF67LCQ:-KWwGLPuXd-8fpY2GciJ:b-pcfmx6) ([raw file](https://github.com/pauloborba/ta/blob/b8dfdf0806966649aea20a09fb7124bd3fe4d00d/test/cucumber/EmailNewGrades.feature#L18)):_

> ``` Cucumber
>     Then the system asks if I want to resend
> 
> #  Cant be tested yet, need importing grades
> ```

não entendi. deveria ter feito um stub

---

_[test/cucumber/steps/EmailSteps.groovy, line 38 at r1](https://reviewable.io:443/reviews/pauloborba/ta/33#-KWwGlHY7nKNHk-s50YW:-KWwGlHY7nKNHk-s50YX:b-klvm9i) ([raw file](https://github.com/pauloborba/ta/blob/b8dfdf0806966649aea20a09fb7124bd3fe4d00d/test/cucumber/steps/EmailSteps.groovy#L38)):_

> ``` Groovy
>         page.selectStudentByEmail(email)
>         at ShowStudentPage
>         assert page.isAllSent()
> ```

não teria como verificar o log do serviço de email?

---

_[test/cucumber/steps/EmailSteps.groovy, line 50 at r1](https://reviewable.io:443/reviews/pauloborba/ta/33#-KWwGwzU8UU33Na3SHQI:-KWwGwzU8UU33Na3SHQJ:b7xvy2t) ([raw file](https://github.com/pauloborba/ta/blob/b8dfdf0806966649aea20a09fb7124bd3fe4d00d/test/cucumber/steps/EmailSteps.groovy#L50)):_

> ``` Groovy
>     to CreateCriterionPage
>     page.createCriterion("C1")
>     to AddEvaluationPage
> ```

merece umas extrações de método

---

_[test/cucumber/steps/EmailSteps.groovy, line 61 at r1](https://reviewable.io:443/reviews/pauloborba/ta/33#-KWwH1MfH13zeNnRot1l:-KWwH1MfH13zeNnRot1m:b-96lyp4) ([raw file](https://github.com/pauloborba/ta/blob/b8dfdf0806966649aea20a09fb7124bd3fe4d00d/test/cucumber/steps/EmailSteps.groovy#L61)):_

> ``` Groovy
> }
> Then(~/^the system asks if I want to resend$/) { ->
>     at ResendPage
> ```

mas como tem certeza que, nessa página, está sendo solicitado o reenvio?

---

_[test/cucumber/steps/EmailSteps.groovy, line 65 at r1](https://reviewable.io:443/reviews/pauloborba/ta/33#-KWwH9kc1u481v8fcFZh:-KWwH9kc1u481v8fcFZi:b3kre) ([raw file](https://github.com/pauloborba/ta/blob/b8dfdf0806966649aea20a09fb7124bd3fe4d00d/test/cucumber/steps/EmailSteps.groovy#L65)):_

> ``` Groovy
> When(~/^I successfully import students evaluation for the "([^"]*)" criterion\.\.\.$/) { String arg1 ->
>     // Write code here that turns the phrase above into concrete actions
>     throw new PendingException()
> ```

?

---

_[test/cucumber/steps/EmailSteps.groovy, line 69 at r1](https://reviewable.io:443/reviews/pauloborba/ta/33#-KWwHAzOBvjN787sPUX2:-KWwHAzOBvjN787sPUX3:b3kre) ([raw file](https://github.com/pauloborba/ta/blob/b8dfdf0806966649aea20a09fb7124bd3fe4d00d/test/cucumber/steps/EmailSteps.groovy#L69)):_

> ``` Groovy
> Then(~/^The system is marked as having grades to send$/) { ->
>     // Write code here that turns the phrase above into concrete actions
>     throw new PendingException()
> ```

?

---

_[test/functional/pages/StudentPages/ResendPage.groovy, line 10 at r1](https://reviewable.io:443/reviews/pauloborba/ta/33#-KW-olY6hFYzGxZa5jjo-r1-10:-KWwHJjGRJkYeTaFCdAp:b-r1zh4z) ([raw file](https://github.com/pauloborba/ta/blob/b8dfdf0806966649aea20a09fb7124bd3fe4d00d/test/functional/pages/StudentPages/ResendPage.groovy#L10)):_
<details><summary>

<i>Previously, rcrs3 (Rodrigo Carneiro) wrote…</i></summary>



> o jeito certo de passar o titulo da página seria utilizando a classe GetPageTitle() para tornar-la independente de linguagem
> </details>
> exato!

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/pauloborba/ta/33)_

**Comment by pauloborba** [comment]
Reviewed 2 of 8 files at r2.
Review status: 11 of 17 files reviewed at latest revision, 16 unresolved discussions.

---

*[grails-app/controllers/ta/MailerController.groovy, line 1 at r2](https://reviewable.io:443/reviews/pauloborba/ta/33#-KXzmel3yJ56TO7fnNr4:-KXzmel3yJ56TO7fnNr5:b138x77) ([raw file](https://github.com/pauloborba/ta/blob/de0bfac264787ee78879cd411814fc9fde88c4b1/grails-app/controllers/ta/MailerController.groovy#L1)):*
> ```Groovy
> package ta
> ```

deveria ser um serviço, não um controlador

---

*[grails-app/domain/ta/EvaluationsByCriterion.groovy, line 10 at r2](https://reviewable.io:443/reviews/pauloborba/ta/33#-KXzq9yTxSay_gDhuWcK:-KXzq9yTxSay_gDhuWcL:b678023) ([raw file](https://github.com/pauloborba/ta/blob/de0bfac264787ee78879cd411814fc9fde88c4b1/grails-app/domain/ta/EvaluationsByCriterion.groovy#L10)):*
> ```Groovy
>     static hasMany = [evaluations:Evaluation]
>     double criterionAverage
>     boolean PendingMail
> ```

renomeia com primeira letra minúscula

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pauloborba/ta/33)*

File: grails-app/domain/ta/EvaluationsByCriterion.groovy [hunk 7]
```diff
@@ -41,15 +41,19 @@ class EvaluationsByCriterion {
...
-        if(s.length()==0)return ""
-        s=s.substring(0,s.length()-1)
-        return ""+criterion.getDescription()+":"+s
+        return FormatMessage(s)
+    }
+
+    private String FormatMessage(String s) {
+        if (s.length() == 0) return ""
+        s = s.substring(0, s.length() - 1)
+        return " *" + criterion.getDescription() + ": " + s + "\n"
```
Comment (rodCarneiro) [inline grails-app/domain/ta/EvaluationsByCriterion.groovy:56]: Seria melhor criar a string primeiro e depois retornar ela

File: test/cucumber/steps/EmailSteps.groovy [hunk 8]
```diff
@@ -59,12 +59,4 @@ And(~/^I am on the students index page$/) { ->
 }
 Then(~/^the system asks if I want to resend$/) { ->
     at ResendPage
```
Comment (rodCarneiro) [inline test/cucumber/steps/EmailSteps.groovy:61]: Continua sem ter certeza que o sistema está solicitando o reenvio

**Pull request closed without merging**